core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -128,33 +130,40 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
                                       networkClient: NetworkClient,
                                       txnStateManager: TransactionStateManager,
                                       txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
-                                      time: Time,
-                                      metrics:Metrics) extends Logging {
-
-  private def addQueueLengthMetric(sensor: Sensor, namePrefix: String, description: String): Unit = {
-    sensor.add(metrics.metricName(s"$namePrefix-queue-length",
-      "transaction-marker-channel-metrics",
-      description,
-      Collections.singletonMap("broker-id",config.brokerId.toString)),
-      new Total())
-  }
-
-  private val markersQueuePerBroker: concurrent.Map[Int, TxnMarkerQueue] = concurrent.TrieMap.empty[Int, TxnMarkerQueue]
-
-  private val markersQueueForUnknownBroker = new TxnMarkerQueue(Node.noNode)
-  private val unknownBrokerQueueSensor = metrics.sensor("unknown-broker-queue")
-  addQueueLengthMetric(unknownBrokerQueueSensor, "unknown-broker", "the number of WriteTxnMarker requests with unknown brokers")
+                                      time: Time) extends Logging with KafkaMetricsGroup {
 
   private val interBrokerListenerName: ListenerName = config.interBrokerListenerName
 
   private val txnMarkerSendThread: InterBrokerSendThread = {
     new InterBrokerSendThread("TxnMarkerSenderThread-" + config.brokerId, networkClient, drainQueuedTransactionMarkers, time)
   }
 
+  private val markersQueuePerBroker: concurrent.Map[Int, TxnMarkerQueue] = concurrent.TrieMap.empty[Int, TxnMarkerQueue]
+
+  private val markersQueueForUnknownBroker = new TxnMarkerQueue(Node.noNode)
   private val txnLogAppendRetryQueue = new LinkedBlockingQueue[TxnLogAppend]()
 
-  private val txnLogAppendRetryQueueSensor = metrics.sensor("txn-log-append-retry-queue")
-  addQueueLengthMetric(txnLogAppendRetryQueueSensor, "txn-log-append-retry", "the number of txn log appends that need to be retried")
+  // we use separate atomic integers since the size() call on the above two lists are expensive
```

Comment:
For `TxnMarkerQueue` I add the integer outside since I only want to maintain the counter for the `unknown` queue but not for other queues. But thinking about it a bit more I feel this optimization may not worth it: so okay I will move it inside as well.